runtime/src/stakes.rs

```diff
@@ -39,46 +43,92 @@ impl Stakes {
     pub fn history(&self) -> &StakeHistory {
         &self.stake_history
     }
-    pub fn clone_with_epoch(&self, next_epoch: Epoch) -> Self {
+
+    pub fn activate_epoch(
+        &mut self,
+        next_epoch: Epoch,
+        optimize_epoch_boundary_updates: bool,
+        thread_pool: &ThreadPool,
+    ) {
         let prev_epoch = self.epoch;
-        if prev_epoch == next_epoch {
-            self.clone()
-        } else {
+        self.epoch = next_epoch;
+
+        thread_pool.install(|| {
+            let stake_delegations = &self.stake_delegations;
+            let stake_history = &mut self.stake_history;
+            let vote_accounts: &VoteAccountsHashMap = self.vote_accounts.as_ref();
+
+            // construct map of vote pubkey -> list of stake delegations
+            let vote_delegations: HashMap<Pubkey, Vec<&Delegation>> = {
+                let mut vote_delegations = HashMap::with_capacity(vote_accounts.len());
+                stake_delegations
+                    .iter()
+                    .for_each(|(_stake_pubkey, delegation)| {
+                        let vote_pubkey = &delegation.voter_pubkey;
+                        vote_delegations
+                            .entry(*vote_pubkey)
+                            .and_modify(|delegations: &mut Vec<_>| delegations.push(delegation))
+                            .or_insert_with(|| vec![delegation]);
+                    });
+                vote_delegations
+            };
+
```

Comment:
Don't think this is a big deal ~200ms, but the validators that upgraded without the feature flag enabled will be doing this extra bit of work compared to the old `new_stake_history_entry()` code right?

Reply:
Yes, exactly. I was also thinking that it wasn't a big deal and leaving it like this made it easier to keep the feature switched optimization code clean